### ISSUE TITLE
StackOverflowError when trying to inject annotated class into an Interceptor

### ISSUE THREAD

**Issue by eserban** [payara/Payara#1459]
# Description #
----------

Injecting a simple CDI bean into an Interceptor and having the same CDI bean annotated with the interceptor's annotation causes a StackOverflowError.

## Expected Outcome

This injection (even if the described scenario sounds a bit weird) was working on Glassfish 3.1, which leads me to think that this might be a bug.

As far as I know Interceptor classes have the same lifecycle as their associated target class, so I guess that when an intercepted bean is instantiated the interceptor is also instantiated, but since my interceptor need the same bean, a new one is created, which will create a new interceptor and so on. However, even if I use a Singleton bean, the result is the same -> StackOverflowError.

## Current Outcome

```
[Payara 4.1] [WARNING] [] [org.glassfish.jersey.internal.Errors] [levelValue: 900] [[
  The following warnings have been detected: WARNING: Unknown HK2 failure detected:
MultiException stack 1 of 1
java.lang.StackOverflowError
	at org.jboss.weld.injection.producer.AbstractInstantiator.newInstance(AbstractInstantiator.java:28)
	at org.jboss.weld.injection.producer.BasicInjectionTarget.produce(BasicInjectionTarget.java:112)
	at org.jboss.weld.injection.producer.BeanInjectionTarget.produce(BeanInjectionTarget.java:180)
	at org.jboss.weld.bean.ManagedBean.create(ManagedBean.java:158)
	at org.jboss.weld.context.unbound.DependentContextImpl.get(DependentContextImpl.java:70)
	at org.jboss.weld.bean.ContextualInstanceStrategy$DefaultContextualInstanceStrategy.get(ContextualInstanceStrategy.java:100)
	at org.jboss.weld.bean.ContextualInstance.get(ContextualInstance.java:50)
	at org.jboss.weld.manager.BeanManagerImpl.getReference(BeanManagerImpl.java:744)
	at org.jboss.weld.bean.interceptor.CdiInterceptorFactory.create(CdiInterceptorFactory.java:46)
	at org.jboss.weld.interceptor.proxy.InterceptionContext.initInterceptorInstanceMap(InterceptionContext.java:103)
	at org.jboss.weld.interceptor.proxy.InterceptionContext.of(InterceptionContext.java:78)
	at org.jboss.weld.interceptor.proxy.InterceptionContext.forNonConstructorInterception(InterceptionContext.java:74)
	at org.jboss.weld.injection.producer.InterceptorApplyingInstantiator.newInstance(InterceptorApplyingInstantiator.java:60)
	at org.jboss.weld.injection.producer.BasicInjectionTarget.produce(BasicInjectionTarget.java:112)
	at org.jboss.weld.injection.producer.BeanInjectionTarget.produce(BeanInjectionTarget.java:180)
	at org.jboss.weld.bean.ManagedBean.create(ManagedBean.java:158)
	at org.jboss.weld.context.unbound.DependentContextImpl.get(DependentContextImpl.java:70)
	at org.jboss.weld.bean.ContextualInstanceStrategy$DefaultContextualInstanceStrategy.get(ContextualInstanceStrategy.java:100)
	at org.jboss.weld.bean.ContextualInstance.get(ContextualInstance.java:50)
	at org.jboss.weld.manager.BeanManagerImpl.getReference(BeanManagerImpl.java:744)
	at org.jboss.weld.manager.BeanManagerImpl.getInjectableReference(BeanManagerImpl.java:844)
	at org.jboss.weld.injection.FieldInjectionPoint.inject(FieldInjectionPoint.java:92)
	at org.jboss.weld.util.Beans.injectBoundFields(Beans.java:362)
	at org.jboss.weld.util.Beans.injectFieldsAndInitializers(Beans.java:373)
	at org.jboss.weld.injection.producer.ResourceInjector$1.proceed(ResourceInjector.java:70)
	at org.glassfish.weld.services.InjectionServicesImpl.aroundInject(InjectionServicesImpl.java:173)
	at org.jboss.weld.injection.InjectionContextImpl.run(InjectionContextImpl.java:46)
	at org.jboss.weld.injection.producer.ResourceInjector.inject(ResourceInjector.java:72)
	at org.jboss.weld.injection.producer.BasicInjectionTarget.inject(BasicInjectionTarget.java:117)
	at org.jboss.weld.bean.ManagedBean.create(ManagedBean.java:159)
	at org.jboss.weld.context.unbound.DependentContextImpl.get(DependentContextImpl.java:70)
	at org.jboss.weld.bean.ContextualInstanceStrategy$DefaultContextualInstanceStrategy.get(ContextualInstanceStrategy.java:100)
	at org.jboss.weld.bean.ContextualInstance.get(ContextualInstance.java:50)
	at org.jboss.weld.manager.BeanManagerImpl.getReference(BeanManagerImpl.java:744)
	at org.jboss.weld.bean.interceptor.CdiInterceptorFactory.create(CdiInterceptorFactory.java:46)
	at org.jboss.weld.interceptor.proxy.InterceptionContext.initInterceptorInstanceMap(InterceptionContext.java:103)
	at org.jboss.weld.interceptor.proxy.InterceptionContext.of(InterceptionContext.java:78)
	at org.jboss.weld.interceptor.proxy.InterceptionContext.forNonConstructorInterception(InterceptionContext.java:74)
	at org.jboss.weld.injection.producer.InterceptorApplyingInstantiator.newInstance(InterceptorApplyingInstantiator.java:60)
	at org.jboss.weld.injection.producer.BasicInjectionTarget.produce(BasicInjectionTarget.java:112)
	at org.jboss.weld.injection.producer.BeanInjectionTarget.produce(BeanInjectionTarget.java:180)
	at org.jboss.weld.bean.ManagedBean.create(ManagedBean.java:158)
	at org.jboss.weld.context.unbound.DependentContextImpl.get(DependentContextImpl.java:70)
	at org.jboss.weld.bean.ContextualInstanceStrategy$DefaultContextualInstanceStrategy.get(ContextualInstanceStrategy.java:100)
	at org.jboss.weld.bean.ContextualInstance.get(ContextualInstance.java:50)
	at org.jboss.weld.manager.BeanManagerImpl.getReference(BeanManagerImpl.java:744)
	at org.jboss.weld.manager.BeanManagerImpl.getInjectableReference(BeanManagerImpl.java:844)
	at org.jboss.weld.injection.FieldInjectionPoint.inject(FieldInjectionPoint.java:92)
	at org.jboss.weld.util.Beans.injectBoundFields(Beans.java:362)
	at org.jboss.weld.util.Beans.injectFieldsAndInitializers(Beans.java:373)
	at org.jboss.weld.injection.producer.ResourceInjector$1.proceed(ResourceInjector.java:70)
	at org.glassfish.weld.services.InjectionServicesImpl.aroundInject(InjectionServicesImpl.java:173)
	at org.jboss.weld.injection.InjectionContextImpl.run(InjectionContextImpl.java:46)
	at org.jboss.weld.injection.producer.ResourceInjector.inject(ResourceInjector.java:72)
	at org.jboss.weld.injection.producer.BasicInjectionTarget.inject(BasicInjectionTarget.java:117)
	at org.jboss.weld.bean.ManagedBean.create(ManagedBean.java:159)
	at org.jboss.weld.context.unbound.DependentContextImpl.get(DependentContextImpl.java:70)
	at org.jboss.weld.bean.ContextualInstanceStrategy$DefaultContextualInstanceStrategy.get(ContextualInstanceStrategy.java:100)
	at org.jboss.weld.bean.ContextualInstance.get(ContextualInstance.java:50)
	at org.jboss.weld.manager.BeanManagerImpl.getReference(BeanManagerImpl.java:744)
	at org.jboss.weld.bean.interceptor.CdiInterceptorFactory.create(CdiInterceptorFactory.java:46)
	at org.jboss.weld.interceptor.proxy.InterceptionContext.initInterceptorInstanceMap(InterceptionContext.java:103)
	at org.jboss.weld.interceptor.proxy.InterceptionContext.of(InterceptionContext.java:78)
	at org.jboss.weld.interceptor.proxy.InterceptionContext.forNonConstructorInterception(InterceptionContext.java:74)
	at org.jboss.weld.injection.producer.InterceptorApplyingInstantiator.newInstance(InterceptorApplyingInstantiator.java:60)
	at org.jboss.weld.injection.producer.BasicInjectionTarget.produce(BasicInjectionTarget.java:112)
	at org.jboss.weld.injection.producer.BeanInjectionTarget.produce(BeanInjectionTarget.java:180)
	at org.jboss.weld.bean.ManagedBean.create(ManagedBean.java:158)
	at org.jboss.weld.context.unbound.DependentContextImpl.get(DependentContextImpl.java:70)
	at org.jboss.weld.bean.ContextualInstanceStrategy$DefaultContextualInstanceStrategy.get(ContextualInstanceStrategy.java:100)
	at org.jboss.weld.bean.ContextualInstance.get(ContextualInstance.java:50)
	at org.jboss.weld.manager.BeanManagerImpl.getReference(BeanManagerImpl.java:744)
	at org.jboss.weld.manager.BeanManagerImpl.getInjectableReference(BeanManagerImpl.java:844)
	at org.jboss.weld.injection.FieldInjectionPoint.inject(FieldInjectionPoint.java:92)
	at org.jboss.weld.util.Beans.injectBoundFields(Beans.java:362)
	at org.jboss.weld.util.Beans.injectFieldsAndInitializers(Beans.java:373)
	at org.jboss.weld.injection.producer.ResourceInjector$1.proceed(ResourceInjector.java:70)
	at org.glassfish.weld.services.InjectionServicesImpl.aroundInject(InjectionServicesImpl.java:173)
```

## Steps to reproduce

Simple web application with a web.xml/Application class, a REST resource, a bean and an interceptor.

```
<web-app xmlns="http://java.sun.com/xml/ns/javaee"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/web-app_2_5.xsd"
         metadata-complete="false"
         version="2.5">

    <display-name>Test</display-name>
</web-app>
```

```
@ApplicationPath("/")
public class ApplicationConfig extends Application {
    public Set<Class<?>> getClasses() {
        return new HashSet<Class<?>>(singletonList(TestResource.class));
    }
}
```

```
@InterceptorBinding
@Target({TYPE, METHOD})
@Retention(RUNTIME)
public @interface Intercept {
}
```

```
@Interceptor
@Intercept
public class TestInterceptor {

    @Inject
    private TestService testService;

    @AroundInvoke
    public Object logMethodEntry(InvocationContext ctx) throws Exception {
        testService.method();
        return ctx.proceed();
    }
}
```

```
@Path("/test")
public class TestResource {

    //injected here only to make sure the TestService bean is instantiated at least once
    @Inject
    private TestService service;

    @GET
    @Produces(MediaType.APPLICATION_JSON)
    public String message() {
        return service.method();
    }
}
```

```
public class TestService {

    @Intercept //without this annotation the problem does not reproduce
    public String method() {
        return "success";
    }
}
```

1 -**  Start a domain & deploy war application
2 -**  Call GET http://<host>:<port>/<app path>/test to trigger the injection.

## Environment ##

- **Payara Version**: 4.1.1.171
- **Edition**: Full
- **JDK Version**: 8 Oracle
- **Operating System**: Windows

**Comment by fturizo** [comment]
Greetings @eserban,

I reproduced the problem quickly on both _GlassFish 3.1.2.2_ and _Payara 4.1.1.171_, however, there are some changes needed for the CDI artefacts to be correctly configured:

1. The TestService has no scope definitions at all. By default, if a CDI bean lacks a scope definition, it's treated with the **dependent** scope. 
2. Likewise, the `TestResource` has no scope declared, so the server doesn't recognise it as a CDI bean. Deploying it like that causes a `NullPointerException`.

By simply annotating these components with `@RequestScoped` I can deploy the application and make the interceptor work on _GlassFish 3.1.2.2_, however on Payara I get another **StackOverflowError**:

```log
Warning: StandardWrapperValve[fish.payara.support.samples.loopinterceptor.ApplicationConfig]: Servlet.service() for servlet fish.payara.support.samples.loopinterceptor.ApplicationConfig threw exception
java.lang.StackOverflowError
	at org.apache.felix.framework.BundleImpl.isExtension(BundleImpl.java:884)
	at org.apache.felix.framework.BundleImpl.loadClass(BundleImpl.java:917)
	at org.jvnet.hk2.osgiadapter.OSGiModuleImpl$4$1.run(OSGiModuleImpl.java:432)
	at org.jvnet.hk2.osgiadapter.OSGiModuleImpl$4$1.run(OSGiModuleImpl.java:429)
	at java.security.AccessController.doPrivileged(Native Method)
	at org.jvnet.hk2.osgiadapter.OSGiModuleImpl$4.loadClass(OSGiModuleImpl.java:429)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at com.sun.enterprise.v3.server.APIClassLoaderServiceImpl$APIClassLoader.loadClass(APIClassLoaderServiceImpl.java:243)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:411)
	at com.sun.enterprise.loader.CurrentBeforeParentClassLoader.loadClass(CurrentBeforeParentClassLoader.java:83)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:411)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at org.glassfish.web.loader.WebappClassLoader.loadClass(WebappClassLoader.java:1766)
....
```
This is actually a side effect of a change introduced on the _CDI specification 1.1_. Prior to it, interceptor beans could have a declared scope and by default were treated as being **application scoped**. However, on [CDI 1.1](http://docs.jboss.org/cdi/spec/1.1/cdi-spec.html#builtin_scopes) which is shipped with Payara 4.1.x.x:

> 2.4.1. Built-in scope types
> 
> ...
> 
> **If an interceptor or decorator has any scope other than @Dependent, non-portable behaviour results.**

On Weld, all interceptor beans are treated as having dependent scopes (as mentioned by the spec), which causes a circular dependency instantiation loop, overflowing the method stack.

Side effect or not, I wouldn't code a circular dependency between your service and interceptor. I encourage you to separate the behaviour of the injected service needed for the interceptor to work into a separate service, or prevent that particular business method from being intercepted. That'd be the suggested practice.

**Comment by fturizo** [comment]
Closing this issue since we haven't received a reply from the user yet and the issue has been clarified.
